ZuluContent/Zulu/Engines/Classes/SkillCheck.cs

```diff
@@ -0,0 +1,255 @@
+                }
+
+                skill.BaseFixedPoint -= amount * 10;
+                return;
+            }
+
+            points = (int) (points * GetSkillPointsMultiplier(from as PlayerMobile, skillName));
+
+            // TODO: Should we implement global multipliers?
+
+            AwardPoints(from as PlayerMobile, skillName, points);
+        }
+
+        public static double GetSkillPointsMultiplier(PlayerMobile from, SkillName skillName)
+        {
+            if (from.ZuluClass.IsSkillInClass(skillName))
+                return from.ZuluClass.Bonus;
+            return 1.0;
+        }
+
+        public static void AwardPoints(PlayerMobile from, SkillName skillName, int points)
+        {
+            var config = Configs[skillName];
+
+            Skill skill = from.Skills[skillName];
+
+            AwardRawAttributePoints(from, skill, points);
+
+            // Check the strength advancement
+            if (config.StrAdvancement != null && Utility.Random(1000) < config.StrAdvancement.Chance * 10)
+            {
+                AwardRawAttributePoints(from, StatType.Str,
+                    Utility.Dice(config.StrAdvancement.PointsAmount, config.StrAdvancement.PointsSides,
+                        config.StrAdvancement.PointsBonus));
+            }
+
+            // Check the dexterity advancement
+            if (config.DexAdvancement != null && Utility.Random(1000) < config.DexAdvancement.Chance * 10)
+            {
+                AwardRawAttributePoints(from, StatType.Dex,
+                    Utility.Dice(config.DexAdvancement.PointsAmount, config.DexAdvancement.PointsSides,
+                        config.DexAdvancement.PointsBonus));
+            }
+
+            // Check intelligence advancement
+            if (config.IntAdvancement != null && Utility.Random(1000) < config.IntAdvancement.Chance * 10)
+            {
+                AwardRawAttributePoints(from, StatType.Int,
+                    Utility.Dice(config.IntAdvancement.PointsAmount, config.IntAdvancement.PointsSides,
+                        config.IntAdvancement.PointsBonus));
+            }
+        }
+
+        public static void AwardRawAttributePoints(PlayerMobile from, Skill skill, double rawPoints)
+        {
+            if (rawPoints == 0.0)
+                return;
+
+            var initial = (int) (skill.Base * 10.0);
+            var current = initial;
+
+            var toIncrease = GetNeededRawPointsToIncrease(current);
+
+            while (rawPoints >= toIncrease)
+            {
+                rawPoints -= toIncrease;
+                ++current;
+                toIncrease = GetNeededRawPointsToIncrease(current);
+            }
+
+            if (Utility.Random((int) toIncrease) < rawPoints)
+                ++current;
+
+            if (current != initial)
+            {
+                if (current > 60000)
+                    current = 60000;
+
+                // TODO: Add cmd level check
+                if (current > 1350)
```

Comment:
Wondering where a good place would be to set the skill/stat caps?

Reply:
That stat cap is only used directly by the SkillCheck code we are replacing and it looks like the caps set inside the setters are really high values, so we probably don't need to worry about changing anything else